Mathlib/MeasureTheory/SetSemiring.lean

```diff
@@ -0,0 +1,278 @@
+    intro u hu x hx
+    exact hC.diffFinset_subset hu ht hx
+  have hJu_disj : ∀ (u) (hu : u ∈ C), (Ju u hu : Set (Set α)).PairwiseDisjoint id := fun u hu ↦
+    hC.pairwiseDisjoint_diffFinset hu ht
+  have hJu_sUnion : ∀ (u) (hu : u ∈ C), ⋃₀ (Ju u hu : Set (Set α)) = u \ t :=
+    fun u hu ↦ hC.sUnion_diffFinset hu ht
+  have hJu_disj' : ∀ (u) (hu : u ∈ C) (v) (hv : v ∈ C) (_h_dis : Disjoint u v),
+      Disjoint (⋃₀ (Ju u hu : Set (Set α))) (⋃₀ ↑(Ju v hv)) :=by
+    intro u hu v hv huv_disj
+    rw [hJu_sUnion, hJu_sUnion]
+    exact disjoint_of_subset (Set.diff_subset u t) (Set.diff_subset v t) huv_disj
+  let J' : Finset (Set α) := Finset.biUnion (Finset.univ : Finset J) fun u ↦ Ju u (h_ss u.prop)
+  have hJ'_subset : ↑J' ⊆ C := by
+    intro u
+    simp only [Subtype.coe_mk, univ_eq_attach, coe_biUnion, mem_coe, mem_attach, iUnion_true,
+      mem_iUnion, Finset.exists_coe, bex_imp]
+    intro v hv huvt
+    exact hJu_subset v (h_ss hv) huvt
+  refine ⟨J', hJ'_subset, ?_, ?_⟩
+  · rw [Finset.coe_biUnion]
+    refine PairwiseDisjoint.biUnion ?_ ?_
+    · simp only [univ_eq_attach, mem_coe, id.def, iSup_eq_iUnion]
+      simp_rw [PairwiseDisjoint, Set.Pairwise, Function.onFun]
+      intro x _ y _ hxy
+      have hxy_disj : Disjoint (x : Set α) y := by
+        by_contra h_contra
+        refine hxy ?_
+        refine Subtype.ext ?_
+        exact h_dis.elim x.prop y.prop h_contra
+      convert hJu_disj' (x : Set α) (h_ss x.prop) y (h_ss y.prop) hxy_disj
+      · rw [sUnion_eq_biUnion]
+        congr
+      · rw [sUnion_eq_biUnion]
+        congr
+    · exact fun u _ ↦ hJu_disj _ _
+  · rw [coe_insert, sUnion_insert, Set.union_comm, ← Set.diff_diff, h_eq]
+    simp_rw [sUnion_eq_biUnion, Set.iUnion_diff]
+    simp only [Subtype.coe_mk, mem_coe, Finset.mem_biUnion, Finset.mem_univ, exists_true_left,
+      Finset.exists_coe, iUnion_exists, true_and]
+    rw [iUnion_comm]
+    refine iUnion_congr fun i ↦ ?_
+    by_cases hi : i ∈ J
+    · simp only [hi, iUnion_true, exists_prop]
+      rw [← hJu_sUnion i (h_ss hi), sUnion_eq_biUnion]
+      simp only [mem_coe]
+    · simp only [hi, iUnion_of_empty, iUnion_empty]
+
+open Classical in
+/-- In a semiring of sets `C`, for all set `s ∈ C` and finite set of sets `I ⊆ C`,
+`diffFinset₀` is a finite set of sets in `C` such that `s \ ⋃₀ I = ⋃₀ (hC.diffFinset₀ hs I hI)`.
+`diffFinset` is as a special case of `diffFinset₀` where `I` is a singleton. -/
+noncomputable def diffFinset₀ (hC : IsSetSemiring C) (hs : s ∈ C) (hI : ↑I ⊆ C) : Finset (Set α) :=
+  (hC.exists_disjoint_finset_diff_eq hs hI).choose \ {∅}
+
+lemma empty_not_mem_diffFinset₀ (hC : IsSetSemiring C) (hs : s ∈ C) (hI : ↑I ⊆ C) :
+    ∅ ∉ hC.diffFinset₀ hs hI := by
+  classical
+  simp only [diffFinset₀, mem_sdiff, Finset.mem_singleton, eq_self_iff_true, not_true,
+    and_false_iff, not_false_iff]
+
+lemma diffFinset₀_subset (hC : IsSetSemiring C) (hs : s ∈ C) (hI : ↑I ⊆ C) :
+    ↑(hC.diffFinset₀ hs hI) ⊆ C := by
```

Comment:
I would hope the coercion gets inserted automatically?
```suggestion
    hC.diffFinset₀ hs hI ⊆ C := by
```

Reply:
It isn't inserted automatically.